scrapy/utils/curl.py

```diff
@@ -0,0 +1,82 @@
+import argparse
+from shlex import split
+
+from six.moves.http_cookies import SimpleCookie
+from six.moves.urllib.parse import urlparse
+from six import string_types, iteritems
+from w3lib.http import basic_auth_header
+
+
+class CurlParser(argparse.ArgumentParser):
+    def error(self, message):
+        error_msg = \
+            'There was an error parsing the curl command: {}'.format(message)
+        raise ValueError(error_msg)
+
+
+curl_parser = CurlParser()
+curl_parser.add_argument('url')
+curl_parser.add_argument('-H', '--header', dest='headers', action='append')
+curl_parser.add_argument('-X', '--request', dest='method', default='get')
+curl_parser.add_argument('-d', '--data', dest='data')
+curl_parser.add_argument('-u', '--user', dest='auth')
+
+curl_parser.add_argument('--compressed', action='store_true')
+curl_parser.add_argument('-s', '--silent', action='store_true')
+
+
+def curl_to_request_kwargs(curl_args):
+    """Convert a curl command syntax to Request kwargs
+
+    :param curl_args: string containing the curl command
+    :return: dictionary of Request kwargs
+    """
+
+    if isinstance(curl_args, string_types):
+        curl_args = split(curl_args)
+
+    if curl_args[0] != 'curl':
+        raise ValueError('A curl command must start with "curl"')
+
+    parsed_args, argv = curl_parser.parse_known_args(curl_args[1:])
+    if argv:
+        msg = 'Unrecognized arguments: %s'
+        raise ValueError(msg % (argv,))
+
+    url = parsed_args.url
+
+    # curl automatically prepends 'http' if the scheme is missing, but Request
+    # needs the scheme to work
+    parsed_url = urlparse(url)
+    if not parsed_url.scheme:
+        url = 'http://'+url
+
+    result = {
+        'method': parsed_args.method.upper(),
+        'url': url,
+    }
+
+    headers = []
+    cookies = {}
+    for header in parsed_args.headers or ():
+        name, val = header.split(':', 1)
+        name = name.strip().title()
```

Comment:
why is .title() needed?

Reply:
My memory is foggy, since it was a while ago, so take it with a pinch of salt, IIRC it provided some case-sensitivity consistency. 

Headers are case-insensitive according to spec, and with JS code being all over the place as it is in the wild, I have seen a lot of cases with curl commands having all-lowercase headers. For those, when you would try to check or override them you would use the more canonical `Title-Cased` version and it would return nothing or end up having both titlecase and lowercase versions at the same time.

I suppose this is a non-issue now that the preferred public API goes through creating a `Request` object that wraps headers in a `CaselessDict`, so we can probably drop this.